AsyncDisplayKit/ASDisplayNode+Yoga.mm

```diff
@@ -0,0 +1,377 @@
+
+  child.yogaParent = self;
+  YGNodeInsertChild(_yogaNode, child.yogaNode, YGNodeGetChildCount(_yogaNode));
+}
+
+- (void)removeYogaChild:(ASDisplayNode *)child
+{
+  if (child == nil) {
+    return;
+  }
+  child.yogaParent = nil;
+  NSInteger index = [_yogaChildren indexOfObjectIdenticalTo:child];
+  if (index != NSNotFound) {
+    [_yogaChildren removeObjectAtIndex:index];
+    YGNodeRemoveChild(_yogaNode, child.yogaNode);
+  }
+  if (_yogaChildren.count == 0 && self.yogaParent == nil) {
+    self.hierarchyState &= ~ASHierarchyStateYogaLayoutEnabled;
+  }
+}
+
+- (ASLayout *)layoutTreeForYogaNode
+{
+  uint32_t childCount = YGNodeGetChildCount(_yogaNode);
+  ASDisplayNodeAssert(childCount == self.yogaChildren.count,
+                      @"Yoga tree should always be in sync with .yogaNodes array! %@", self.yogaChildren);
+
+  NSMutableArray *sublayouts = [NSMutableArray arrayWithCapacity:childCount];
+  for (ASDisplayNode *subnode in self.yogaChildren) {
+    [sublayouts addObject:[subnode layoutTreeForYogaNode]];
+  }
+
+  CGSize  size     = CGSizeMake(YGNodeLayoutGetWidth(_yogaNode), YGNodeLayoutGetHeight(_yogaNode));
+  CGPoint position = CGPointMake(YGNodeLayoutGetLeft(_yogaNode), YGNodeLayoutGetTop(_yogaNode));
+
+  ASDisplayNodeLogEvent(self, @"Yoga calculatedSize: %@", NSStringFromCGSize(size));
+
+  ASLayout *layout = [ASLayout layoutWithLayoutElement:self
+                                                  size:size
+                                              position:position
+                                            sublayouts:sublayouts];
+  return layout;
+}
+
+- (void)setYogaMeasureFuncIfNeeded
+{
+  // Manual size calculation via calculateSizeThatFits:
+  // This will be used for ASTextNode, as well as any other leaf node that has no layout spec.
+  if ((self.methodOverrides & ASDisplayNodeMethodOverrideLayoutSpecThatFits) == NO
+      && self.layoutSpecBlock == NULL &&  self.yogaChildren.count == 0) {
+    YGNodeSetContext(_yogaNode, (__bridge void *)self);
+    YGNodeSetMeasureFunc(_yogaNode, &ASLayoutElementYogaMeasureFunc);
+  }
+}
+
+- (ASLayout *)calculateLayoutFromYogaRoot:(ASSizeRange)rootConstrainedSize
+{
+  if (ASHierarchyStateIncludesYogaLayoutMeasuring(self.hierarchyState)) {
+    ASDisplayNodeAssert(NO, @"A Yoga layout is being performed by a parent; children must not perform their own until it is done! %@", [self displayNodeRecursiveDescription]);
+    return [ASLayout layoutWithLayoutElement:self size:CGSizeZero];
+  }
+
+  ASDisplayNodePerformBlockOnEveryYogaChild(self, ^(ASDisplayNode * _Nonnull node) {
+    node.hierarchyState |= ASHierarchyStateYogaLayoutMeasuring;
+  });
+
+  YGNodeRef rootYogaNode = self.yogaNode;
+
+  // Apply the constrainedSize as a base, known frame of reference.
+  // If the root node also has style.*Size set, these will be overridden below.
+  YGNodeStyleSetMinWidth (rootYogaNode, yogaFloatForCGFloat(rootConstrainedSize.min.width));
```

Comment:
I have planned changes to how this copying of style information should work.  However, I'd like to get the initial version in before changing this much more.  The reason is that I can't test against a large Yoga layout without doing a repository sync, and syncing is tough right now because of divergence from master.  In any case, consider it my responsibility to make sure this logic becomes cleaner, even though it is hidden in the +Yoga file.